File: osint_nexus/core/detectors/cdn.py
from collections.abc import Mapping
from beartype import beartype

from osint_nexus.core.detectors.base import FingerprintStrategy
from osint_nexus.core.type_defs import JSONValue


class CdnFingerprintStrategy(FingerprintStrategy[Mapping[str, JSONValue], Mapping[str, JSONValue]]):
    """Strategy for CDN Header fingerprinting."""

    name: str = "cdn_headers"

    @beartype
    def extract(self, data: Mapping[str, JSONValue]) -> Mapping[str, JSONValue]:
        # Expecting data: {"server_headers": dict[str, str]}
        headers_raw = data.get("server_headers")
        headers = headers_raw if isinstance(headers_raw, Mapping) else {}

        # Identify common CDN headers
        cdn_identified = "cf-ray" in headers or "x-amz-cf-id" in headers or "x-cache" in headers

        fingerprint: Mapping[str, JSONValue] = {
            "cdn_detected": cdn_identified,
            "provider": headers.get("server"),
        }

        confidence = 0.75 if cdn_identified else 0.1

        return {
            "name": self.name,
            "data": fingerprint,
            "confidence": confidence,
        }

File: internal/engine/strategies/cdn.go
package strategies

import (
	"context"
	"strings"

	"github.com/osint-nexus/internal/types"
)

// CDNPayload implements types.FingerprintPayload for CDN fingerprinting input.
type CDNPayload struct {
	ServerHeaders map[string]string `json:"server_headers"`
}

// PayloadType returns the payload type identifier.
func (p CDNPayload) PayloadType() string { return "cdn_input" }

// CDNOutputPayload implements types.FingerprintPayload for CDN fingerprinting output.
type CDNOutputPayload struct {
	CdnDetected bool   `json:"cdn_detected"`
	Provider    string `json:"provider"`
}

// PayloadType returns the payload type identifier.
func (p CDNOutputPayload) PayloadType() string { return "cdn_output" }

// CdnFingerprintStrategy identifies CDN presence based on HTTP headers.
type CdnFingerprintStrategy struct{}

// NewCdnFingerprintStrategy initializes a new CdnFingerprintStrategy.
func NewCdnFingerprintStrategy() *CdnFingerprintStrategy {
	return &CdnFingerprintStrategy{}
}

// Name returns the strategy identifier.
func (s *CdnFingerprintStrategy) Name() string {
	return "cdn_headers"
}

// Extract analyzes HTTP headers to detect CDN presence.
func (s *CdnFingerprintStrategy) Extract(ctx context.Context, data types.FingerprintData) (types.FingerprintResult, error) {
	payload, ok := data.Payload.(CDNPayload)
	if !ok {
		return types.FingerprintResult{
			Name: s.Name(),
			Data: types.FingerprintData{
				Type:    "cdn_detection",
				Payload: CDNOutputPayload{CdnDetected: false, Provider: ""},
			},
			Confidence: 0.1,
		}, nil
	}

	// Identify common CDN headers
	cdnIdentified := false
	for k := range payload.ServerHeaders {
		lowerKey := strings.ToLower(k)
		if lowerKey == "cf-ray" || lowerKey == "x-amz-cf-id" || lowerKey == "x-cache" {
			cdnIdentified = true
			break
		}
	}

	confidence := 0.1
	if cdnIdentified {
		confidence = 0.75
	}

	serverHeader := ""
	for k, v := range payload.ServerHeaders {
		if strings.ToLower(k) == "server" {
			serverHeader = v
			break
		}
	}

	return types.FingerprintResult{
		Name: s.Name(),
		Data: types.FingerprintData{
			Type:    "cdn_detection",
			Payload: CDNOutputPayload{CdnDetected: cdnIdentified, Provider: serverHeader},
		},
		Confidence: confidence,
	}, nil
}
